src/sentry/models/dynamicsampling.py

```diff
@@ -0,0 +1,258 @@
+import hashlib
+from datetime import datetime, timedelta
+from typing import Any, Mapping, Optional, Sequence
+
+from django.db import connections, models, router, transaction
+from django.db.models import Q
+from django.utils import timezone
+
+from sentry.backup.scopes import RelocationScope
+from sentry.db.models import FlexibleForeignKey, Model, region_silo_only_model
+from sentry.utils import json
+
+# max number of custom rules that can be created per organization
+MAX_CUSTOM_RULES = 2000
+CUSTOM_RULE_START = 3000
+
+
+class TooManyRules(ValueError):
+    """
+    Raised when a there is already the max number of rules active for an organization
+    """
+
+    pass
+
+
+def get_condition_hash(condition: Any) -> str:
+    """
+    Returns the hash of the rule based on the condition and projects
+    """
+    condition_string = to_order_independent_string(condition)
+    # make it a bit shorter
+    return hashlib.sha1(condition_string.encode("utf-8")).hexdigest()
+
+
+def to_order_independent_string(val: Any) -> str:
+    """
+    Converts a value in an order independent string and then hashes it
+
+    Note: this will insure the same repr is generated for ['x', 'y'] and ['y', 'x']
+        Also the same repr is generated for {'x': 1, 'y': 2} and {'y': 2, 'x': 1}
+    """
+    ret_val = ""
+    if isinstance(val, Mapping):
+        for key in sorted(val.keys()):
+            ret_val += f"{key}:{to_order_independent_string(val[key])}-"
+    elif isinstance(val, (list, tuple)):
+        for item in sorted(val):
+            ret_val += f"{to_order_independent_string(item)}-"
+    else:
+        ret_val = str(val)
+    return ret_val
+
+
+@region_silo_only_model
+class CustomDynamicSamplingRuleProject(Model):
+    """
+    Many-to-many relationship between a custom dynamic sampling rule and a project.
+    """
+
+    __relocation_scope__ = RelocationScope.Organization
+
+    custom_dynamic_sampling_rule = FlexibleForeignKey(
+        "sentry.CustomDynamicSamplingRule", on_delete=models.CASCADE
+    )
+    project = FlexibleForeignKey("sentry.Project", on_delete=models.CASCADE)
+    organization = FlexibleForeignKey("sentry.Organization", on_delete=models.CASCADE)
```

Comment:
Why do you have a foreign key to the organization table here?

Reply:
Can't you directly infer it from the main table?

Reply:
It was suggested by @jan-auer to simplify cleanup jobs.